src/hardware/disney.cpp

```diff
@@ -19,456 +20,214 @@
-		if(disney.stereo) {
-			uint8_t gapfiller0 = 128;
-			uint8_t gapfiller1 = 128;
-			if (real_used) {
-				gapfiller0 = disney.da[0].buffer[real_used-1];
-				gapfiller1 = disney.da[1].buffer[real_used-1];
-			};
-
-			memset(disney.da[0].buffer+real_used,
-				gapfiller0,len-real_used);
-			memset(disney.da[1].buffer+real_used,
-				gapfiller1,len-real_used);
-
-			DISNEY_PlayStereo(len, disney.da[0].buffer, disney.da[1].buffer);
-			len -= real_used;
-
-		} else { // mono
-			uint8_t gapfiller = 128; // Keep the middle
-			if (real_used) {
-				// fix for some stupid game; it outputs 0 at the end of the stream
-				// causing a click. So if we have at least two bytes availible in the
-				// buffer and the last one is a 0 then ignore that.
-				if(disney.leader->buffer[real_used-1]==0)
-					real_used--;
-			}
-			// do it this way because AddSilence sounds like a gnawing mouse
-			if (real_used)
-				gapfiller = disney.leader->buffer[real_used-1];
-			//LOG_MSG("gapfiller %x, fill len %d, realused %d",gapfiller,len-real_used,real_used);
-			memset(disney.leader->buffer+real_used,	gapfiller, len-real_used);
-			disney.chan->AddSamples_m8(len, disney.leader->buffer);
-		}
-		disney.da[0].used =0;
-		disney.da[1].used =0;
-
-		//LOG_MSG("disney underflow %d",len - real_used);
+	// First, add any frames we've queued since the last callback
+	while (frames_remaining && render_queue.size()) {
+		channel->AddSamples_sfloat(1, &render_queue.front()[0]);
+		render_queue.pop();
+		--frames_remaining;
 	}
-	if (disney.last_used+100<PIC_Ticks) {
-		// disable sound output
-		PIC_AddEvent(DISNEY_disable, 0.0001); // I think we shouldn't delete the
-		                                      // mixer while we are inside it
+	// If the queue's run dry, render the remainder and sync-up our time datum
+	while (frames_remaining) {
+		const auto frame = Render();
+		channel->AddSamples_sfloat(1, &frame[0]);
+		--frames_remaining;
 	}
+	last_rendered_ms = PIC_FullIndex();
 }
 
-static void DISNEY_ShutDown([[maybe_unused]] Section *sec)
+std::unique_ptr<Disney> disney = {};
+
+Disney::Disney(const std::string_view filter_pref)
 {
-	DEBUG_LOG_MSG("DISNEY: Shutting down");
+	// Prime the FIFO with a single silent sample
+	fifo.emplace(Mixer_GetSilentDOSSample<uint8_t>());
 
-	// Remove interrupt events
-	PIC_RemoveEvents(DISNEY_disable);
+	using namespace std::placeholders;
+	const auto audio_callback = std::bind(&Disney::AudioCallback, this, _1);
 
-	// Stop the game from accessing the IO ports
-	disney.read_handler.Uninstall();
-	disney.write_handler.Uninstall();
+	// Setup the mixer callback
+	channel = MIXER_AddChannel(audio_callback,
+	                           dac_rate_hz,
+	                           "DISNEY",
+	                           {ChannelFeature::Sleep,
+	                            ChannelFeature::ReverbSend,
+	                            ChannelFeature::ChorusSend,
+	                            ChannelFeature::DigitalAudio});
+	assert(channel);
```

Comment:
You'll need to initialise the channel with the mixer sample rate so we can the zero-order-hold upsampling to emulate the crunchiness, so just use 0 instead of `dac_rate_hz`.